### PR TITLE
Prevent location state changes during emergency calls

### PR DESCRIPTION
**Pull request by LaakkonenJussi**
AML is ran with the emergency calls and it requires to keep location
services, i.e., GPS on during the call. It must not be allowed for the
user to try to change this during the call.

This change tracks the call state with QMceCallState funcionality. When
state is "Active" and type is "Emergency" all changes made to the
location or GPS status are prevented and warning is logged.

### CODE REVIEW

File: src/locationsettings.cpp [hunk 1]
```diff
@@ -814,3 +839,17 @@ void LocationSettingsPrivate::writeSettings()
         }
     }
 }
+
+bool LocationSettingsPrivate::isEmergencyCallActive()
+{
+    return m_callstate == QMceCallState::Active && m_calltype == QMceCallState::Emergency;
```
Comment (spiiroin) [inline src/locationsettings.cpp:845]: Type can be "emergency" only if/when there is active (or incoming) emergency call, so in theory one could ignore state, test just the type (and get those theoretical incoming emergency calls handled for free).

File: src/locationsettings.cpp [hunk 2]
```diff
@@ -466,6 +479,12 @@ bool LocationSettings::locationEnabled() const
 void LocationSettings::setLocationEnabled(bool enabled)
 {
     Q_D(LocationSettings);
+
+    if (d->isEmergencyCallActive()) {
+        qWarning() << "Cannot allow to change location enabled status when emergency call is active";
```
Comment (spiiroin) [inline src/locationsettings.cpp:484]: Just wondering whether we want debug logging for any attempt to enable/disable, or just for the ones that would actually lead to changes i.e. should this be withing the below "if it differs from current" conditional block?

Also, is it okay for qml property setter not to set the property? I'd guess yes, but does somebody know for sure?

Comment (spiiroin) [reply src/locationsettings.cpp:484]: Sort of related, if there were properties like "canToggleLocationEnabled" or stte -> those could perhaps be used in (qml) ui to disable/hide those element that could lead to attempts to make such changes?

**Review by spiiroin — APPROVED** [review]
Logic wise lgtm.
Some behavioral bits of which I'm not sure if they matter or not.

File: src/locationsettings.cpp [hunk 3]
```diff
@@ -814,3 +839,17 @@ void LocationSettingsPrivate::writeSettings()
...
+    return m_callstate == QMceCallState::Active && m_calltype == QMceCallState::Emergency;
+}
+
+void LocationSettingsPrivate::onCallStateChanged()
+{
+    if (!m_mceCallState->valid())
+        return;
+
+    m_callstate = m_mceCallState->state();
+    m_calltype = m_mceCallState->type();
```
Comment (pvuorela) [inline src/locationsettings.cpp:854]: Any reason to have member variables for these instead of just calling getters when evaluating the state?

File: src/locationsettings.cpp [hunk 4]
```diff
@@ -466,6 +479,12 @@ bool LocationSettings::locationEnabled() const
 void LocationSettings::setLocationEnabled(bool enabled)
 {
     Q_D(LocationSettings);
+
+    if (d->isEmergencyCallActive()) {
```
Comment (pvuorela) [inline src/locationsettings.cpp:483]: Pondering should it be allowed to turn the location/gps on.

Comment (LaakkonenJussi) [reply src/locationsettings.cpp:483]: Why? AML does it already.

File: src/locationsettings.cpp (hunk 3 above)
Comment (LaakkonenJussi) [reply src/locationsettings.cpp:854]: No, no any particular reason. I can drop those if polling is preferred.

File: src/locationsettings.cpp (hunk 4 above)
Comment (LaakkonenJussi) [reply src/locationsettings.cpp:483]: And AML controls the location anyways, when battery is low it keeps running but shuts down all power consumers to have some grace period for the user to find a charger. Anyways, it would compete with AML requirements and the AML service, which will force the value to the file causing maybe confusion as in such case user would see "-> on -> off" and vice versa. Better to have this ignore all cases. 

Or maybe if we could have an additional QMceCallState or variable `aml_enabled` ? In such case all actions would be blocked but when AML is not enabled user could control this.

File: src/locationsettings.cpp (hunk 2 above)
Comment (LaakkonenJussi) [reply src/locationsettings.cpp:484]: That is a good option I think, to have a `canToggleLocation` visible upwards to UI.

File: src/locationsettings.cpp (hunk 3 above)
Comment (pvuorela) [reply src/locationsettings.cpp:854]: Looks like on mce-qt these are in the end plain "return iValue;" type of code so the extra copies shouldn't provide much. -> Let's just call the getters when the status is needed.

Comment (LaakkonenJussi) [reply src/locationsettings.cpp:854]: Yeah I was just looking the same that it does not create a new request but just returns internal value. On it.

File: src/locationsettings.cpp (hunk 4 above)
Comment (pvuorela) [reply src/locationsettings.cpp:483]: Thinking of robustness and limiting only the minimum. AML is sort of an add-on so these rules on this level are not totally unproblematic, and also it's now assuming that AML will not use this component while on the other hand it would be good if all access to the file would go via this.

File: src/locationsettings.cpp (hunk 3 above)
Comment (LaakkonenJussi) [reply src/locationsettings.cpp:854]: But, if we'd need the `canToggleGPS` property I think we'd need to keep track of the state in order for that to work.

Comment (pvuorela) [reply src/locationsettings.cpp:854]: In that case it could boil down to boolean. Not sure which would be better but such canToggle & protection could be even implemented fully in the UI side as the location setting modification is couple very specific places. 

Speaking of that, there's also already the policy support for disallowing location settings changes in the UI.

Comment (LaakkonenJussi) [reply src/locationsettings.cpp:854]: An `enum` vs. boolean? I'd take the former any day in state tracking. It is more future proof. I think the safeguard should not be only in UI but could be also there to track what is going on beneath.

Where is that policy support implemented?

Comment (pvuorela) [reply src/locationsettings.cpp:854]: Boolean as in canToggleGPS sounds quite much like such.

Policy is sailfish-policy, living with the mdm libraries. UI side honors the policy, e.g. systemsettings / location.qml, disables the settings if the policy prevents usage.

File: src/locationsettings.cpp (hunk 1 above)
Comment (spiiroin) [reply src/locationsettings.cpp:845]: Think type as substate of ongoing call. When there is no call i.e. state=="none" , type is dontcare (but shows up as "normal"). When there is an active call, type can be "emergency" based on what ofono reports for call/modem. Technically there is nothing stopping from incoming call being labeled as emergency too - but I do not recall if that is something that can actually happen. But if it can....

Checking for "active and emergency" is true when call to emergency number is initiated / call from emergency number has been answered.

Whereas "(active or incoming) and emergency" == "emergency" is true like above + also when call from emergency number is ringing but has not been answered yet.

It is fine as it is and I guess which way to test depends on how it needs to behave  -> I'll leave it up to you whether these obscure corner case details matter or not ;-)

Comment (LaakkonenJussi) [reply src/locationsettings.cpp]: Yep, thanks for clarifying this.

File: src/locationsettings.cpp (hunk 2 above)
Comment (LaakkonenJussi) [reply src/locationsettings.cpp:483]: Added some with 517f7bd029dc2e9cc22e1b2323b63db32d0a3659 -> naming needs to be rethought. On Friday evening brain just shortcircuits.

File: src/locationsettings.cpp (hunk 3 above)
Comment (LaakkonenJussi) [reply src/locationsettings.cpp]: I really think the `enum` serves here as we might need more granularity 517f7bd029dc2e9cc22e1b2323b63db32d0a3659 (naming improvement suggestions are welcome).

File: src/locationsettings.cpp (hunk 4 above)
Comment (LaakkonenJussi) [reply src/locationsettings.cpp]: AML is a service that at some point should also be launched at the lock UI so I think this is not available then. It is simple enough just to prevent user access to the options as there is a legislative requirement to do so. I left opening for enable only option in 517f7bd029dc2e9cc22e1b2323b63db32d0a3659 which might be needed. Anyways, all emergency calls != AML is on.

Comment (spiiroin) [reply src/locationsettings.cpp]: I'm starting to think that if the situation is not yet clear enough to have a future-proof mw policy it might be better not to implement such a thing. Especially if the only gotcha would be having device user repeatedly and persistently toggling things in settings app during an ongoing emergency call. Or was there something else too?

**Comment by LaakkonenJussi** [comment]
In general I think the `org.sailfishos.aml` service name is the best to monitor here as that is only registered when AML is active for the region the emergency call is made in. But this is to be split off from the JB#57710 and will be continued in another task somewhere in the future.
